ping/ping.c

```diff
@@ -1466,13 +1477,82 @@ int ping4_send_probe(struct ping_rts *rts, socket_st *sock, void *packet,
 	if (rts->timing && !rts->opt_latency) {
 		struct timeval tmp_tv;
 		gettimeofday(&tmp_tv, NULL);
-		memcpy(icp + 1, &tmp_tv, sizeof(tmp_tv));
+		if (rts->probe == 0)
+			memcpy(icp + 1, &tmp_tv, sizeof(tmp_tv));
+		else if (rts->probe == 1) {
+			rts->timestamp_offset = sizeof(struct icmphdr) + sizeof(struct exthdr) + ntohs(iio.len);
+			memcpy((char *)icp + rts->timestamp_offset, &tmp_tv, sizeof(tmp_tv));
+		}
 		icp->checksum = in_cksum((unsigned short *)&tmp_tv, sizeof(tmp_tv), ~icp->checksum);
 	}
-
 	i = sendto(sock->fd, icp, cc, 0, (struct sockaddr *)&rts->whereto, sizeof(rts->whereto));
 
 	return (cc == i ? 0 : i);
+build_probe:
+	extbase = (struct exthdr *)(icp + 1);
+	iiobase = (struct iiohdr *)((char *)extbase + sizeof(struct exthdr));
+	icp->type = ICMP_EXT_ECHO;
+	/* PROBE messages use only the first 8 bits as sequence number */
+	icp->un.echo.sequence = htons((rts->ntransmitted + 1) << 8);
+	icp->un.echo.sequence |= htons(1);	/* Set L-bit */
+	WRITE_VERSION(ext.v_rsvd , 2);
+	ext.v_rsvd = htons(ext.v_rsvd);
+	ext.checksum = 0;
```

Comment:
You must fill in the checksum later. While it appears that the Linux kernel does not check the checksum, there are other probe implementations that do.

My tcpdump update https://github.com/the-tcpdump-group/tcpdump/pull/1115 will check that it's filled in properly.